fxrunner/src/lib/proto.rs

```diff
@@ -106,44 +127,190 @@ where
                 .expect("unzip task was cancelled or panicked");
 
                 if let Err(e) = unzip_result {
-                    self.inner
-                        .send(DownloadBuildReply {
-                            result: Err(e.into_error_message()),
-                        })
-                        .await?;
+                    self.send(DownloadBuildReply {
+                        result: Err(e.into_error_message()),
+                    })
+                    .await?;
 
                     Err(e.into())
                 } else {
                     let firefox_path = download_dir.join("firefox").join("firefox.exe");
 
                     if !firefox_path.exists() {
                         let err = RunnerProtoError::MissingFirefox;
-                        self.inner
-                            .send(DownloadBuildReply {
-                                result: Err(err.into_error_message()),
-                            })
-                            .await?;
+                        self.send(DownloadBuildReply {
+                            result: Err(err.into_error_message()),
+                        })
+                        .await?;
 
                         Err(err)
                     } else {
-                        self.inner
-                            .send(DownloadBuildReply {
-                                result: Ok(DownloadStatus::Extracted),
-                            })
-                            .await?;
+                        self.send(DownloadBuildReply {
+                            result: Ok(DownloadStatus::Extracted),
+                        })
+                        .await?;
 
                         Ok(firefox_path)
                     }
                 }
             }
 
             Err(e) => {
-                error!(self.log, "could not download build"; "error" => %e);
-                self.inner
-                    .send(DownloadBuildReply {
+                error!(self.log, "could not download build"; "error" => ?e);
+                self.send(DownloadBuildReply {
+                    result: Err(e.into_error_message()),
+                })
+                .await?;
+                Err(e.into())
+            }
+        }
+    }
+
+    pub async fn send_profile_reply(
+        &mut self,
+        download_dir: &Path,
+    ) -> Result<Option<PathBuf>, RunnerProtoError<S::Error>> {
+        let SendProfile { profile_size } = self.recv().await?;
+
+        match profile_size {
+            None => {
+                self.send(SendProfileReply { result: Ok(None) }).await?;
+
+                Ok(None)
+            }
+
+            Some(profile_size) => {
```

Comment:
There are a few other long match arms in this patch, which might benefit from a similar refactoring, but it might just be a matter of individual preference here 😅